native/asb/src/main/java/io/ballerinax/asb/MessageDispatcher.java

```diff
@@ -18,208 +18,310 @@
 
     /**
      * Initialize the Message Dispatcher.
      *
-     * @param service          Ballerina service instance.
-     * @param runtime          Ballerina runtime instance.
-     * @param iMessageReceiver Asb MessageReceiver instance.
+     * @param service       Ballerina service instance.
+     * @param runtime       Ballerina runtime instance.
+     * @param clientBuilder Asb MessageReceiver instance.
      */
-    public MessageDispatcher(BObject service, BObject caller, Runtime runtime, IMessageReceiver iMessageReceiver) {
+    public MessageDispatcher(BObject service, BObject caller, Runtime runtime,
+            ServiceBusProcessorClientBuilder clientBuilder) {
         this.service = service;
         this.caller = caller;
-        this.queueName = iMessageReceiver.getEntityPath();
         this.runtime = runtime;
-        this.receiver = iMessageReceiver;
+        this.builder = clientBuilder;
+        log.setLevel(Level.ERROR);
     }
 
     /**
-     * Start receiving messages asynchronously and dispatch the messages to the attached service.
+     * Start receiving messages asynchronously and dispatch the messages to the
+     * attached service.
      *
      * @param listener Ballerina listener object.
      * @return IllegalArgumentException if failed.
      */
-    public void receiveMessages(BObject listener) {
+    public void receiveMessages(BObject listener, MessageDispatcher md) {
         try {
-            ExecutorService executorService = Executors.newSingleThreadExecutor();
-            this.pumpMessage(receiver, executorService);
+            CountDownLatch countdownLatch = new CountDownLatch(1);
+            ServiceBusProcessorClient processorClient = this.builder.processMessage(t -> {
+                try {
+                    md.processMessage(t);
+                } catch (IOException e) {
+                    log.error("IOException occurred when processing the message", e);
+                    throw new RuntimeException(e);
+                }
+            }).processError(context -> {
+                try {
+                    processError(context, countdownLatch);
+                } catch (IOException e) {
+                    log.error("IOException while processing the error found when processing the message", e);
+                }
+            }).buildProcessorClient();
+            processorClient.start();
+            log.debug("(Message Dispatcher)Receiving started identifier: " + processorClient.getIdentifier());
+            countdownLatch.await();
+            if (processorClient.isRunning()) {
+                processorClient.close();
+                log.debug("(Message Dispatcher)Processor client closed identifier: " + processorClient.getIdentifier());
+            }
+            ;
         } catch (IllegalArgumentException e) {
             ASBUtils.returnErrorValue(e.getMessage());
+        } catch (Exception e) {
+            ASBUtils.returnErrorValue(e.getMessage());
```

Comment:
We might need to improve error handling and sub typing, which we will do after receiving some feedbacks.

Reply:
+1, yes, lets do it after feedbacks